concurrency/SparkJobManagerLocalImpl.java

```diff
@@ -0,0 +1,210 @@
+        }
+
+        System.setProperty("org.apache.logging.log4j.simplelog.StatusLogger.level", "TRACE");
+        RunSparkClassMain runner = RunSparkClassMain.builder()
+                .arguments(args)
+                .classToExecute(classToExecute)
+                .build();
+        Thread t = new Thread(runner);
+        Arn sparkJobArn = Arn.forResource(customerId, Arn.Prefix.SPARK_JOB, "local/" + Long.valueOf(t.getId()).toString());
+        runner.setArn(sparkJobArn);
+        jobMap.put(t.getId(), t);
+        t.start();
+        return sparkJobArn;
+    }
+
+    /**
+     * Stop Spark Job
+     * @param credentials AWS Credentials
+     * @param sparkJobArn Spark Job Arn
+     * @return
+     */
+    public void stopSparkJob(AWSCredentialsProvider credentials, Arn sparkJobArn) {
+        final String customerId = sparkJobArn.getCustomerId();
+        if (sparkJobArn.getCustomerId().compareToIgnoreCase(customerId) != 0) {
+            throw new InvalidArgumentException("Invalid customer ID " + customerId + " for Arn " + sparkJobArn.toString());
+        }
+        SparkJobStatus status = getJobStatus(credentials, sparkJobArn);
+        if (status == SparkJobStatus .ACTIVE) {
+            // We don't really have a good way to stop it...
+            String[] split = sparkJobArn.getResourceId().split("/");
+            Long threadId = Long.valueOf(split[split.length - 1]);
+            Thread thread = jobMap.get(threadId);
+            if (thread != null) {
+                thread.interrupt();
+                jobMap.remove(threadId);
+            }
+        }
+    }
+
+    /**
+     * Get accumulated total Spark Job times for metering.
+     * @param credentials AWS Credentials
+     * @param sparkJobArn Spark Job Arn
+     * @return totalJobTime in milliseconds
+     */
+    @Override
+    public long getTotalJobTime(AWSCredentialsProvider credentials, Arn sparkJobArn) {
+        return -1;
+    }
+
+    /**
+     * Get Spark Job Status
+     * @param credentials AWS Credentials
+     * @param sparkJobArn Spark Job Arn
+     * @return
+     */
+    @Override
+    public SparkJobStatus getJobStatus(AWSCredentialsProvider credentials, Arn sparkJobArn) {
+        final String customerId = sparkJobArn.getCustomerId();
+        if (sparkJobArn.getCustomerId().compareToIgnoreCase(customerId) != 0) {
+            throw new InvalidArgumentException("Invalid customer ID " + customerId + " for Arn " + sparkJobArn.toString());
+        }
+        String[] split = sparkJobArn.getResourceId().split("/");
+        if (split.length > 0) {
+            try {
+                Long threadId = Long.valueOf(split[split.length - 1]);
+                if (jobMap.containsKey(threadId)) {
+                    return SparkJobStatus.ACTIVE;
+                }
+                if (resultMap.containsKey(sparkJobArn.toString())) {
+                    return resultMap.get(sparkJobArn.toString()) ? SparkJobStatus.COMPLETED : SparkJobStatus.FAILED;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **177 and 178**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **197**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).